### PR TITLE
[Meghana-step 16]

### PR DESCRIPTION
**Pull request by Meghana-Reddy2**
DESCRIPTION 

Introduce User

- Let's create a user model
- Let's create the first user with seed
- Let's link users and tasks
- Add an index to the relationship
- Introduce a mechanism to avoid the N+1 problem
- Let's introduce a bullet that automatically detects N+1 queries

DONE

- created a user model
- created the first user with seed
- link users and tasks
- introduced bullet config that automatically detects N+1 queries
- added system & model tests for user

Screenshot 2024-03-06 at 13 29 15
Screenshot 2024-03-06 at 13 47 27

### CODE REVIEW

File: myapp/tmp/cache/bootsnap/compile-cache-iseq/47/460cfc8614a76f
Comment (wataru0) [inline myapp/tmp/cache/bootsnap/compile-cache-iseq/47/460cfc8614a76f:1]: The `tmp/cache` directory is where temporary cache files are stored, so it is better to use gitignore or git/info/exclude to not manage by git.

File: myapp/spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    it 'create user successfully' do
+      user = build(:user)
+      expect(user).to be_valid
+    end
+
+    it 'no error messages when user created successfully' do
+      user = build(:user)
+      user.valid?
+      expect(user.errors).to be_empty
+    end
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb]: If first one success, doesn't the second one also success?
I think either one is fine.

File: myapp/spec/models/user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    it 'user cannot be created without username' do
+      user = build(:user, username: '')
+      expect(user).to be_invalid
+    end
+
+    it 'show error messages if username is empty' do
+      user = build(:user, username: '')
+      user.valid?
+      expect(user.errors[:username]).to eq ["can't be blank"]
+    end
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb]: how about using `context`?
such as
```
context 'when username is empty' do
  before do
    user = build(:user, username: '')
  end

  it 'cannot be created' do
  end

  it 'shows error message' do
  end
end
```

Comment (wataru0) [reply myapp/spec/models/user_spec.rb]: And this helps us to reduce repetition `build(:user, username: '')`.

File: myapp/spec/rails_helper.rb [hunk 3]
```diff
@@ -73,6 +73,17 @@
...
+if Bullet.enable?
+  config.before(:each) do
+    Bullet.start_request
+  end
+
+  config.after(:each) do
+    Bullet.perform_out_of_channel_notifications if Bullet.notification?
+    Bullet.end_request
+  end
+end
```
Comment (wataru0) [inline myapp/spec/rails_helper.rb]: Checks for inconsistent indentation.
https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Layout/IndentationConsistency

**Review by wataru0 — COMMENTED** [review]
good job 👍 I leave some comments.

File: myapp/tmp/cache/bootsnap/compile-cache-iseq/47/460cfc8614a76f
Comment (Meghana-Reddy2) [reply myapp/tmp/cache/bootsnap/compile-cache-iseq/47/460cfc8614a76f:1]: Done. I removed that file

File: myapp/spec/models/user_spec.rb (hunk 1 above)
Comment (Meghana-Reddy2) [reply myapp/spec/models/user_spec.rb]: Yes, both are the same hence removed the second one 🙇‍♀️

File: myapp/spec/models/user_spec.rb (hunk 2 above)
Comment (Meghana-Reddy2) [reply myapp/spec/models/user_spec.rb]: Changed the format . Thank you.

File: myapp/spec/rails_helper.rb (hunk 3 above)
Comment (Meghana-Reddy2) [reply myapp/spec/rails_helper.rb]: I have changed the format of this please check it once. 🙇‍♀️

File: myapp/spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'validaton' do
+    it 'create user successfully' do
+      user = build(:user)
+      expect(user).to be_valid
+    end
+
+    context 'when username is empty' do
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb:13]: [FYI]
As I said "How about using context?", each person has his/her own preference and there is no one good way to write this.
Both those who nest with `context` and those who write tests all on the same hierarchy with `it` have their good points. 🙏 

I will share one reference on how to write a spec. 
https://www.betterspecs.org/

File: myapp/spec/models/user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      before do
+        @user = build(:user, username: '')
+      end
+    
+      it 'user cannot be created without username' do
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb]: I think `it does not create user` is more natural in a sentence than `it user cannot be created` wdyt? 👀

File: myapp/spec/models/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        expect(@user).to be_invalid
+      end
+    
+      it 'show error messages if username is empty' do
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb]:
```suggestion
      it 'shows error messages if username is empty' do
```

File: myapp/spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        expect(@user.errors[:username]).to include("can't be blank")
+      end
+    end 
+
+    it 'user cannot be created with username longer than 255 characters' do
+      user = build(:user, username: 'a' * 256)
+      expect(user).to be_invalid
+    end
+
+    it 'show error messages if username is longer than 255 characters' do
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb]:
```suggestion
    it 'shows error messages if username is longer than 255 characters' do
```

File: myapp/spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      user.valid?
+      expect(user.errors[:username]).to include('has already been taken')
+    end
+
+    it 'user cannot be created without password_digest' do
+      user = build(:user, password_digest: nil)
+      expect(user).to be_invalid
+    end
+
+    it 'show error messages if password_digest is empty' do
```
Comment (wataru0) [inline myapp/spec/models/user_spec.rb]:
```suggestion
    it 'shows error messages if password_digest is empty' do
```

File: myapp/spec/rails_helper.rb
Comment (wataru0) [inline myapp/spec/rails_helper.rb:1]: I think still there are some coding style problems. recommend you to use `rubocop spec/rails_helper.rb`
https://rubocop.org/

**Review by wataru0 — COMMENTED** [review]
some comments 👍 Please check it!

File: myapp/spec/rails_helper.rb
Comment (Meghana-Reddy2) [reply myapp/spec/rails_helper.rb:1]: Thank you for this 🙏 .

File: myapp/spec/models/user_spec.rb (hunk 4 above)
Comment (Meghana-Reddy2) [reply myapp/spec/models/user_spec.rb:13]: context seems more organized but I think I am more inclined towards writing just 'it' :😅😥 please check [4b776d3](https://github.com/Fablic/training/pull/1619/commits/4b776d396d9c5005042c47b590da0f68dade2704)  and [3166cfe](https://github.com/Fablic/training/pull/1619/commits/3166cfef3b7c8269458fbcbefa11a3235457d6c4) file and let me know about any further changes 🙇‍♀️

File: myapp/spec/models/user_spec.rb (hunk 5 above)
Comment (Meghana-Reddy2) [reply myapp/spec/models/user_spec.rb]: Yeah, That makes more sense ,will change it. Thank you.

File: myapp/spec/models/user_spec.rb (hunk 4 above)
Comment (wataru0) [reply myapp/spec/models/user_spec.rb:13]: Of course, your feeling is also correct! 
all `it` test is easier to add new test cases 👍 
This is a preference.

**Review by wataru0 — APPROVED** [review]
Looks Good To Me! 👍 
LGTM

**Pull request merged**